library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt

```diff
@@ -638,42 +654,51 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
     @ScaledPan
     private fun checkPanBounds(horizontal: Boolean, allowOverScroll: Boolean): Float {
         @ScaledPan val value = if (horizontal) scaledPanX else scaledPanY
-        val viewSize = if (horizontal) mContainerWidth else mContainerHeight
+        val containerSize = if (horizontal) mContainerWidth else mContainerHeight
         @ScaledPan val contentSize = if (horizontal) mTransformedRect.width() else mTransformedRect.height()
-        @ScaledPan val basePanValue = computeBasePan()[(if (horizontal) 0 else 1)]
 
         val overScrollable = if (horizontal) mOverScrollHorizontal else mOverScrollVertical
         @ScaledPan val overScroll = (if (overScrollable && allowOverScroll) maxOverScroll else 0).toFloat()
-        return getPanCorrection(value, viewSize, contentSize, overScroll, basePanValue)
+        return getPanCorrection(value, containerSize, contentSize, overScroll, horizontal)
     }
 
     @ScaledPan
-    private fun getPanCorrection(@ScaledPan value: Float, viewSize: Float,
+    private fun getPanCorrection(@ScaledPan value: Float, containerSize: Float,
                                  @ScaledPan contentSize: Float, @ScaledPan overScroll: Float,
-                                 @ScaledPan basePanValue: Float): Float {
-        @ScaledPan val tolerance = overScroll.toInt()
+                                 horizontal: Boolean): Float {
         var min: Float
         var max: Float
-        if (contentSize <= viewSize) {
-            when(mSmallerPolicy) {
+        if (contentSize <= containerSize) {
+            // If content is smaller than container, act according to the smaller policy.
+            // Expect the output to be >= 0, we will show part of the container background.
+            val extraSpace = containerSize - contentSize // > 0
+            when (mSmallerPolicy) {
                 ZoomApi.SMALLER_POLICY_FROM_TRANSFORMATION -> {
-                    min = basePanValue
-                    max = basePanValue
-                } else  -> {
-                    // If contentSize <= viewSize, we want to stay centered.
-                    // Need a positive translation, that shows some background.
-                    min = (viewSize - contentSize) / 2f
-                    max = (viewSize - contentSize) / 2f
+                    // Apply the transformation gravity.
+                    val correction = applyGravity(mTransformation, extraSpace, horizontal)
+                    min = correction
+                    max = correction
+                }
+                ZoomApi.SMALLER_POLICY_CENTER  -> {
+                    // Stay centered. Need a positive translation, that shows some background.
+                    min = extraSpace / 2f
+                    max = extraSpace / 2f
+                }
+                ZoomApi.SMALLER_POLICY_NONE -> {
+                    // Everything is fine, just don't exit the container boundaries.
```

Comment:
![](https://media.giphy.com/media/NTur7XlVDUdqM/giphy.gif)

(No problem here, just a random tought while reading the comment :sweat_smile: )